File: server/src/intTest/java/org/tctalent/server/repository/db/EducationLevelRepositoryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedEducationLevel;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.EducationLevel;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class EducationLevelRepositoryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private EducationLevelRepository repo;
  private EducationLevel educationLevel;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());
    educationLevel = getSavedEducationLevel(repo);
  }

  @Test
  public void testFindByStatus() {
    List<EducationLevel> result = repo.findByStatus(Status.active);
    assertNotNull(result);
    assertFalse(result.isEmpty());
  }

  @Test
  public void testFindByStatusSingle() {
    List<EducationLevel> result = repo.findByStatus(Status.deleted);
    assertNotNull(result);
    assertFalse(result.isEmpty());
    List<Long> ids = result.stream().map(EducationLevel::getId).toList();
    assertEquals(educationLevel.getId(), ids.getFirst());
  }

  @Test
  public void testFindByStatusSingleFail() {
    educationLevel.setStatus(Status.active);
    repo.save(educationLevel);
    List<EducationLevel> result = repo.findByStatus(Status.deleted);
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }

  @Test
  public void testFindByNameIgnoreCaseFailStatus() {
    EducationLevel result = repo.findByNameIgnoreCase(educationLevel.getName());
    assertNull(result);
  }

  @Test
  public void testFindByNameIgnoreCase() {
    educationLevel.setStatus(Status.active);
    repo.save(educationLevel);
    EducationLevel result = repo.findByNameIgnoreCase(educationLevel.getName());
    assertNotNull(result);
    assertEquals(educationLevel.getId(), result.getId());
  }

  @Test
  public void testFindByNameIgnoreCaseFail() {
    EducationLevel result = repo.findByNameIgnoreCase("NONE TO FIND");
    assertNull(result);
  }

  @Test
  public void testFindByLevelIgnoreCaseFailLevel() {
    EducationLevel result = repo.findByLevelIgnoreCase(9);
    assertNull(result);
  }

  @Test
  public void testFindByLevelIgnoreCaseFailStatus() {
    EducationLevel result = repo.findByLevelIgnoreCase(7);
    assertNull(result);
  }

  @Test
  public void testFindByLevelIgnoreCase() {
    educationLevel.setStatus(Status.active);
    repo.save(educationLevel);
    EducationLevel result = repo.findByLevelIgnoreCase(9);
    assertNotNull(result);
    assertEquals(educationLevel.getId(), result.getId());
  }

  @Test
  public void testFindAllActive() {
    List<EducationLevel> results = repo.findAllActive();
    assertNotNull(results);
    assertFalse(results.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/EducationLevelRepositoryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.EducationLevel
import org.tctalent.server.model.db.Status
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest
import org.tctalent.server.repository.db.integrationhelp.getSavedEducationLevel

class EducationLevelRepositoryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: EducationLevelRepository
  private lateinit var educationLevel: EducationLevel

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    educationLevel = getSavedEducationLevel(repo)
  }

  @Test
  fun `test find by status`() {
    val result = repo.findByStatus(Status.active)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
  }

  @Test
  fun `test find by status single`() {
    val result = repo.findByStatus(Status.deleted)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    val ids = result.map { it.id }
    assertEquals(educationLevel.id, ids.first())
  }

  @Test
  fun `test find by status single fail`() {
    repo.save(educationLevel.apply { status = Status.active })
    val result = repo.findByStatus(Status.deleted)
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }

  @Test
  fun `test find by name ignore case fail status`() {
    val result = repo.findByNameIgnoreCase(educationLevel.name)
    assertNull(result)
  }

  @Test
  fun `test find by name ignore case`() {
    repo.save(educationLevel.apply { status = Status.active })
    val result = repo.findByNameIgnoreCase(educationLevel.name)
    assertNotNull(result)
    assertEquals(educationLevel.id, result.id)
  }

  @Test
  fun `test find by name ignore case fail`() {
    val result = repo.findByNameIgnoreCase("NONE TO FIND")
    assertNull(result)
  }

  @Test
  fun `test find by level ignore case fail level`() {
    val result = repo.findByLevelIgnoreCase(9)
    assertNull(result)
  }

  @Test
  fun `test find by level ignore case fail status`() {
    val result = repo.findByLevelIgnoreCase(7)
    assertNull(result)
  }

  @Test
  fun `test find by level ignore case`() {
    repo.save(educationLevel.apply { status = Status.active })
    val result = repo.findByLevelIgnoreCase(9)
    assertNotNull(result)
    assertEquals(educationLevel.id, result.id)
  }

  @Test
  fun `test find all active`() {
    val results = repo.findAllActive()
    assertNotNull(results)
    assertTrue { results.isNotEmpty() }
  }
}
